flax/linen/transforms.py

```diff
@@ -476,165 +476,94 @@ def __eq__(self, other):
+        )
+        _check_field_is_hashable((*path, 'scope'), static_scope)
+        fingerprints.append(('scope', static_scope))
+      fingerprint = tuple(fingerprints)
+      return type(obj), fingerprint
   elif dataclasses.is_dataclass(obj):
-    ignore = ('parent', 'name') if isinstance(obj, Module) else ()
+    fingerprints = []
     for field in dataclasses.fields(obj):
-      if field.name in ignore:
+      if not hasattr(obj, field.name):
         continue
       value = getattr(obj, field.name)
-      next_path = (*path, field.name)
-      if not _is_hashable(value):
-        yield next_path, value
-        yield from _recursive_find_unhashable_iter(next_path, value, seen_ids)
-  elif not _is_hashable(obj):
-    yield path, obj
-
-
-def _recursive_find_unhashable(name: str, obj: Any):
-  return (
-      ('/'.join(path), obj)
-      for path, obj in _recursive_find_unhashable_iter((name,), obj, set())
-  )
+      fingerprints.append(_field_fingerprint(field.name, value))
+    return type(obj), tuple(fingerprints)
+  elif isinstance(obj, core.DenyList):
+    return type(obj), _field_fingerprint('deny', obj.deny)
+  elif isinstance(obj, dict): # dict goes before serializable to avoid recursion
+    fingerprint = tuple(_field_fingerprint(k, v) for k, v in obj.items())
+    return fingerprint
+  elif serialization.is_serializable(obj):
+    state = serialization.to_state_dict(obj)
+    fingerprint = _fingerprint_recursive(state, path, seen_modules)
+    return type(obj), fingerprint
+  elif isinstance(obj, Mapping):
+    return tuple(_field_fingerprint(k, v) for k, v in obj.items())
+  elif isinstance(obj, Iterable):
+    return tuple(_field_fingerprint(str(i), v) for i, v in enumerate(obj))
+  else:
+    _check_field_is_hashable(path, obj)
+    return obj
 
 
-def _check_field_is_hashable(name: str, x: Any):
+def _check_field_is_hashable(path: tuple[str, ...], x: Any):
   """Checks if a field is hashable."""
-  if not _is_hashable(x):
-    unhashable_paths = ''.join(
-        f'\n- {path}: {type(value)}'
-        for path, value in _recursive_find_unhashable(name, x)
-    )
-    if dataclasses.is_dataclass(x):
-      if x.__hash__ is None:
-        raise ValueError(
-            f"field '{name}' of type '{type(x)}' is a dataclass but its not"
-            ' hashable, using `dataclass(frozen=True, eq=True)` to make it'
-            ' immutable and hashable, or `dataclass(unsafe_hash=True)` for'
-            f' mutable types. Unhashable fields: {unhashable_paths}'
-        )
-      else:
-        raise ValueError(
-            f"field '{name}' of type '{type(x)}' is a hashable dataclass but"
-            ' hashing failed, this probably means that at least one of its'
-            ' fields is not hashable. Unhashable fields:'
-            f' {unhashable_paths}'
-        )
-    else:
-      raise ValueError(
-          f"type '{x}' is not hashable. Unhashable fields: {unhashable_paths}"
-      )
+  try:
+    hash(x)
```

Comment:
would `isinstance(x, typing.Hashable)` work instead of a try-catch block?

Reply:
Nope. Having a `__hash__` method is not enough.